mesonbuild/mdevenv.py

```diff
@@ -145,7 +152,9 @@ def run(options: argparse.Namespace) -> int:
             args = [shell_env]
         elif is_windows():
             shell = get_windows_shell()
-            if shell in ['powershell.exe', 'pwsh.exe']:
+            if not shell:
+                mlog.warning('Failed to determine Windows shell, fallback to cmd.exe')
```

Comment:
There seem to now be 3 cases here:
- no shell, just use cmd.exe
- shell is known powershell one
- shell is something else entirely, we cannot handle that, just use cmd.exe

But we only warn for the first case, not the third?

Reply:
Not sure Windows has other shells. But if the script cannot detect cmd not PowerShell, it will warn. 1 and 3 are the same.

Reply:
Huh wait, that's right. The ps1 script can only ever Write-Host if it finds one of those three executables.

So uh, in that case... why do we need to check `if shell in POWERSHELL_EXES:`, why not just `if shell`?

Either way, the `if not shell: mlog.warning(...)`  should probably be moved to the "else" clause, 5 lines down.

Reply:
The script can also print `cmd.exe`.

Reply:
The return values of get_windows_shell() are `'cmd.exe'`, `'pwsh.exe'`, `'powershell.exe'`, `''` (script works but did not detect any known shell), `None` (script failed to execute).

Reply:
Ok, so still:

```python
shell = get_windows_shell()
if shell in POWERSHELL_EXES:
    do_powershell_stuff()
else:
    if not shell:
        mlog.warning('Failed to determine Windows shell, fallback to cmd.exe')
    do_cmd_dot_exe_stuff()
```